File: clients/scala/src/main/scala/komus24/model/VehicleType.scala
package komus24.model

import komus24.util.StreamUtil

/**
 * Vehicle type options
 *
 * @param name Name
 * @param radius Radius
 * @param weight Weight
 * @param maxBackwardsSpeed Maximal backwads movement speed
 * @param maxSpeed Maximal forward movement speed
 * @param acceleration Acceleration
 * @param friction Friction coefficient
 * @param maxRotateSpeed Maximal rotation speed
 * @param rotateAcceleration Rotational acceleration
 * @param maxFuel Maximal amount of fuel
 * @param fuelUseSpeed Fuel usage speed
 */
case class VehicleType(name: String, radius: Double, weight: Double, maxBackwardsSpeed: Double, maxSpeed: Double, acceleration: Double, friction: Double, maxRotateSpeed: Double, rotateAcceleration: Double, maxFuel: Double, fuelUseSpeed: Double) {
    /**
     * Write VehicleType to output stream
     */
    def writeTo(stream: java.io.OutputStream): scala.Unit = {
        StreamUtil.writeString(stream, name)
        StreamUtil.writeDouble(stream, radius)
        StreamUtil.writeDouble(stream, weight)
        StreamUtil.writeDouble(stream, maxBackwardsSpeed)
        StreamUtil.writeDouble(stream, maxSpeed)
        StreamUtil.writeDouble(stream, acceleration)
        StreamUtil.writeDouble(stream, friction)
        StreamUtil.writeDouble(stream, maxRotateSpeed)
        StreamUtil.writeDouble(stream, rotateAcceleration)
        StreamUtil.writeDouble(stream, maxFuel)
        StreamUtil.writeDouble(stream, fuelUseSpeed)
    }

    /**
     * Get string representation of VehicleType
     */
    override def toString(): String = {
        var stringBuilder = new StringBuilder("VehicleType { ")
        stringBuilder.append("name: ")
        stringBuilder.append('"' + name + '"')
        stringBuilder.append(", ")
        stringBuilder.append("radius: ")
        stringBuilder.append(radius)
        stringBuilder.append(", ")
        stringBuilder.append("weight: ")
        stringBuilder.append(weight)
        stringBuilder.append(", ")
        stringBuilder.append("maxBackwardsSpeed: ")
        stringBuilder.append(maxBackwardsSpeed)
        stringBuilder.append(", ")
        stringBuilder.append("maxSpeed: ")
        stringBuilder.append(maxSpeed)
        stringBuilder.append(", ")
        stringBuilder.append("acceleration: ")
        stringBuilder.append(acceleration)
        stringBuilder.append(", ")
        stringBuilder.append("friction: ")
        stringBuilder.append(friction)
        stringBuilder.append(", ")
        stringBuilder.append("maxRotateSpeed: ")
        stringBuilder.append(maxRotateSpeed)
        stringBuilder.append(", ")
        stringBuilder.append("rotateAcceleration: ")
        stringBuilder.append(rotateAcceleration)
        stringBuilder.append(", ")
        stringBuilder.append("maxFuel: ")
        stringBuilder.append(maxFuel)
        stringBuilder.append(", ")
        stringBuilder.append("fuelUseSpeed: ")
        stringBuilder.append(fuelUseSpeed)
        stringBuilder.append(" }")
        stringBuilder.toString()
    }
}

object VehicleType {
    /**
     * Read VehicleType from input stream
     */
    def readFrom(stream: java.io.InputStream): VehicleType = VehicleType(
        StreamUtil.readString(stream),
        StreamUtil.readDouble(stream),
        StreamUtil.readDouble(stream),
        StreamUtil.readDouble(stream),
        StreamUtil.readDouble(stream),
        StreamUtil.readDouble(stream),
        StreamUtil.readDouble(stream),
        StreamUtil.readDouble(stream),
        StreamUtil.readDouble(stream),
        StreamUtil.readDouble(stream),
        StreamUtil.readDouble(stream)
    )
}

File: clients/swift/Sources/Komus24/Model/VehicleType.swift
/// Vehicle type options
public struct VehicleType {
    /// Name
    let name: String

    /// Radius
    let radius: Double

    /// Weight
    let weight: Double

    /// Maximal backwads movement speed
    let maxBackwardsSpeed: Double

    /// Maximal forward movement speed
    let maxSpeed: Double

    /// Acceleration
    let acceleration: Double

    /// Friction coefficient
    let friction: Double

    /// Maximal rotation speed
    let maxRotateSpeed: Double

    /// Rotational acceleration
    let rotateAcceleration: Double

    /// Maximal amount of fuel
    let maxFuel: Double

    /// Fuel usage speed
    let fuelUseSpeed: Double

    /// Read VehicleType from input stream
    static func readFrom<S: InputStream>(_ stream: S) -> VehicleType {
        var name: String
        name = stream.readString()
        var radius: Double
        radius = stream.readDouble()
        var weight: Double
        weight = stream.readDouble()
        var maxBackwardsSpeed: Double
        maxBackwardsSpeed = stream.readDouble()
        var maxSpeed: Double
        maxSpeed = stream.readDouble()
        var acceleration: Double
        acceleration = stream.readDouble()
        var friction: Double
        friction = stream.readDouble()
        var maxRotateSpeed: Double
        maxRotateSpeed = stream.readDouble()
        var rotateAcceleration: Double
        rotateAcceleration = stream.readDouble()
        var maxFuel: Double
        maxFuel = stream.readDouble()
        var fuelUseSpeed: Double
        fuelUseSpeed = stream.readDouble()
        return VehicleType(name: name, radius: radius, weight: weight, maxBackwardsSpeed: maxBackwardsSpeed, maxSpeed: maxSpeed, acceleration: acceleration, friction: friction, maxRotateSpeed: maxRotateSpeed, rotateAcceleration: rotateAcceleration, maxFuel: maxFuel, fuelUseSpeed: fuelUseSpeed)
    }

    /// Write VehicleType to output stream
    func writeTo<S: OutputStream>(_ stream: S) {
        stream.writeString(name)
        stream.writeDouble(radius)
        stream.writeDouble(weight)
        stream.writeDouble(maxBackwardsSpeed)
        stream.writeDouble(maxSpeed)
        stream.writeDouble(acceleration)
        stream.writeDouble(friction)
        stream.writeDouble(maxRotateSpeed)
        stream.writeDouble(rotateAcceleration)
        stream.writeDouble(maxFuel)
        stream.writeDouble(fuelUseSpeed)
    }
}